### PR TITLE
🍃 Form Logs 

### PR DESCRIPTION
**Pull request by shuhaib-aot**
# Issue Tracking

JIRA: 
Issue Type: BUG/ FEATURE

# Changes
when we change the form name and mapper version or workflow it will call this API to keep this log in backend 
image (2)

image

# How has the change been tested?

# Screenshots (if applicable)

# Build Success screenshot (Till a CICD pipeline is set up)

# Notes

### CODE REVIEW

**Comment by sreehari-aot** [comment]
Can u add more description on this PR for understanding the context clearly? Is there any use case for fetching individual logs instead of all?

File: forms-flow-api/src/formsflow_api/models/form_logs.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+        form_log.form_id = data["form_id"]
+        form_log.logs = data["logs"]
+        form_log.save()
+        return form_log
+
+    @classmethod
+    def get_form_logs(cls, form_id):
+        """Return the logs against form id."""
+        if form_id:
+            form_logs = cls.query.filter(cls.form_id == form_id).first()
```
Comment (sumesh-aot) [inline forms-flow-api/src/formsflow_api/models/form_logs.py:28]: Please do not use, first() and use one_or_none(), which will make sure data accuracy.

File: forms-flow-api/src/formsflow_api/models/form_logs.py [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+            if form_logs:
```
Comment (sumesh-aot) [inline forms-flow-api/src/formsflow_api/models/form_logs.py:29]: This is if check is not needed, you can just return the result.

File: forms-flow-api/src/formsflow_api/models/form_logs.py [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+        form_log = FormLogs()
+        form_log.form_id = data["form_id"]
+        form_log.logs = data["logs"]
+        form_log.save()
+        return form_log
+
+    @classmethod
+    def get_form_logs(cls, form_id):
+        """Return the logs against form id."""
+        if form_id:
```
Comment (sumesh-aot) [inline forms-flow-api/src/formsflow_api/models/form_logs.py:27]: For better readability, you can just assert the form_if. Normally add if condition if there is a business logic when it is null, in this case form_id is expected and should fail if it not present.

File: forms-flow-api/src/formsflow_api/models/form_logs.py [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+            form_logs = cls.query.filter(cls.form_id == form_id).first()
+            if form_logs:
+                return form_logs
+        return None
+
+    @classmethod
+    def update_form_logs(cls, form_id, data):
+        """Update form logs against form id."""
+        if form_id:
```
Comment (sumesh-aot) [inline forms-flow-api/src/formsflow_api/models/form_logs.py:36]: same as above, assert here

File: forms-flow-api/src/formsflow_api/models/form_logs.py [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+            form_logs = cls.query.filter(cls.form_id == form_id).first()
```
Comment (sumesh-aot) [inline forms-flow-api/src/formsflow_api/models/form_logs.py:37]: assert form_logs too, as it doesn't make sense to return None for an operation which is supposed to fail.

Comment (sumesh-aot) [inline forms-flow-api/src/formsflow_api/models/form_logs.py:37]: Use on_or_none()

File: forms-flow-api/src/formsflow_api/models/form_logs.py [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+            if form_logs:
+                form_logs.logs = [*form_logs.logs, data]
+                cls.commit()
+                return form_logs
+        return None
+
+    @classmethod
+    def delete_form_logs(cls, form_id):
+        """Delete form logs."""
+        if form_id:
```
Comment (sumesh-aot) [inline forms-flow-api/src/formsflow_api/models/form_logs.py:47]: assert here

File: forms-flow-api/src/formsflow_api/resources/application.py [hunk 7]
```diff
@@ -301,10 +301,7 @@ class ApplicationResourcesByIds(Resource):
     @staticmethod
```
Comment (sumesh-aot) [inline forms-flow-api/src/formsflow_api/resources/application.py:301]: Looks like no changes related to this PR on this file, except some format changes which is highly dependant on the IDE. Please revert this.

File: forms-flow-api/src/formsflow_api/resources/form_logs.py [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+)
+from formsflow_api.services.form_logs import FormlogService
+
+API = Namespace("FormLogs", description="Form logs")
+
+
+@cors_preflight("POST,OPTIONS")
+@API.route("", methods=["POST", "OPTIONS"])
+class FormLogResource(Resource):
+    """API TO MANAGE THE FORM LOGS."""
```
Comment (sumesh-aot) [inline forms-flow-api/src/formsflow_api/resources/form_logs.py:23]: Capital letters are considered rude in comment, please follow our existing pattern for function and class doc.

**Review by sumesh-aot — CHANGES_REQUESTED** [review]
Sorry, bit of confusion for me; what I thought the flow would be is,
**Use case 1**
- User decides to update existing version (e.g : formId **001**)
- Python API, creates a new form in formio with previous form data (e.g, formId **002** )
- Then update the formio current version with the updated form data (formId **001**, will have new form definition and **002** will have old definition)
- Add a log in form_logs with previous data (add form log with **002** entry)
This will go on, and **001** will be always the latest version.
For this, what I understood from @abhilash-aot was, UI will make an API call, python API will handle all these rather UI deciding what to do.

**Use Case 2**
- User decides to save as a new version 
- Python API, creates a new form in formio.
- Create a new form process mapper

CC : @abhilash-aot @sreehari-aot

File: forms-flow-api/src/formsflow_api/models/form_logs.py (hunk 1 above)
Comment (shuhaib-aot) [reply forms-flow-api/src/formsflow_api/models/form_logs.py:28]: okay thank you

**Comment by abhilash-aot** [comment]
> 

Yes the 2 use cases you have mentioned is what is expected. Initially @shuhaib-aot is trying to capture the logs of existing forms without versioning and adding versioning later . So initially implementation of logs for current mapper changes is what he targeted. Will go through in details

**Comment by shuhaib-aot** [comment]
> Can u add more description on this PR for understanding the context clearly? Is there any use case for fetching individual logs instead of all? @shuhaib-aot
the forms logs against form data change. like if we change the form name and change the workflow it will save the logs like who change that and what is the change. so we can fetch the form logs against single form id

**Pull request closed without merging**